### ISSUE TITLE
Détection statique des ambiguïtés de remplacements

### ISSUE THREAD

**Issue by mquandalle** [betagouv/mon-entreprise#1298]
Quand plusieurs remplacements sont actifs à la fois, nous en choisissons un arbitrairement (le "premier"). Comme discuté dans #1139, de part la définition même des remplacements il est inévitable que plusieurs régimes d’exceptions distincts soient applicables à la même "règle de base". En revanche on ne veut pas que _dans une situation donnée_ plusieurs remplacements soient *actifs* en même temps, nous obligeant à en choisir un "au hasard".

On affiche actuellement un avertissement dans la console lorsque plusieurs remplacements sont applicables à une référence donnée (ce qui n'est pas en soit un problème du moment qu'on peut déterminer qu'il n'y aura pas d’ambiguïté lors de l'évaluation). On a actuellement 9000 lignes d'avertissements https://github.com/betagouv/mon-entreprise/blob/c2149ac047278e1ac0be3befba4495a72a14275c/tmp, ce qui n'est pas une information exploitable pour un développeur.

L'idée est d'ajouter une vérification statique s'assurant que quelque soit la situation en entrée, on n'aura pas d’ambigüité sur les remplacements lors de l'évaluation. On pourra ainsi supprimer les avertissements dans la console quand 2 remplacements sont applicables à même référence, mais on aura une *erreur* si on a une ambiguïté de remplacement dans une situation donnée.

Je vois deux logiques à ajouter :

- si plusieurs remplacements ont des conditions différentes, mais des valeurs de remplacement identiques ce n'est pas un problème, car si on prend l'un ou l'autre on aura le même résultat. C'est très souvent le cas maintenant que les `rend non applicable` sont implémentés sous la forme d'un `remplace: non`, on peut avoir 2 raisons actives en même temps de désactiver une règle donnée (par exemple, je n'ai pas le droit au taux réduit maladie car je profite déjà du statut JEI et qu'en plus je gagne déjà trop d'argent).
  L'implémentation exacte reste à affiner mais l'idée peut être du regrouper les remplacements dont les valeurs sont identiques:
  ```
  remplacements applicables:
    - si: cond1
      par: non
    - si: cond2
      par: non
  ```
  serait transformé en 
  ```
  remplacements applicables:
    - si: cond1 OU cond2
      par: non
  ```
  ce qui explicite qu'il n'y a pas de risque d’ambiguïté à l'évaluation
* L'autre cas c'est quand on a des conditions différentes mais mutuellement exclusives. Deux conventions collectives peuvent chacune avoir leur propre valeur pour l'indemnisation d'un repas, ce n'est pas un problème car on sait que ces deux conventions collectives ne peuvent pas êtres actives en même temps.
  L'implémentation nécessitera ici sans doute de regrouper les règles dont les conditions portent sur le même variant:
  ```
  remplacements applicables:
    - si: convention collective = 'restos'
      par: 10 €
    - si: convention collective = 'comptables'
      par: 4 €
  ```
  serait transformé en 
  ```
  remplacements applicables:
    - si: convention collective = 'restos' OU convention collective = 'comptables'
      par: 
        - si: convention collective = 'restos'
          alors: 10€
        - sinon
          alors: 4€
  ```
  rendant une nouvelle fois explicite qu'un seul remplacement sera actif à la fois. Notons que cette logique de regrouper 2 remplacements sous une même condition en ajoutant un “OU” n'est pas valable dans le cas général car elle implique de choisir un ordre dans le remplacement. On peut le faire pour les conventions collectives uniquement parce qu'on peut détecter statiquement que 2 conventions collectives ne peuvent pas être actives en même temps et que l'ordre n'importe pas ici, ce qui implique très probablement d'implémenter  #963 (ce qui permettrait d'ailleurs d'utiliser plutôt un mécanisme "correspondance" ici)

Les `remplacements applicables` sont uniquement là pour illustrer la logique de détection d’ambiguïté, l'implémentation pourra être différente.

Enfin voir le sujet connexe sur la transitivité des remplacements #1296, car il est probable qu'il faille la prendre en compte pour cette vérification statique.

**Comment by johangirod** [comment]
Je suis d'accord avec à peu près tout sauf ce point : 

> On pourra ainsi supprimer les avertissements dans la console quand 2 remplacements sont applicables à même référence, mais on aura une erreur si on a une ambiguïté de remplacement dans une situation donnée.

Il restera forcément des cas où il y aura ambiguité statique, mais qui n'arrivera jamais dans l'utilisation. En fait toute la question est de savoir ou placer le curseur. Je pense à une option "lache" qui autorise les remplacements multiples, et un mode "strict" qui transforme tous les warnings en erreur (comme dans n'importe quel compilateur). 

Par ailleurs, on peut imaginer pouvoir enlever des warnings en ajoutant une priorité aux remplacements (nombre entier). On aurait donc : 

- Warning plus pertinents avec une analyse statique plus poussée (comme tu décris dans l'issue)
- Possibilité de corriger un warning avec une priorité
- Possibilité de transformer les warnings en erreur avec un mode strict

**Comment by mquandalle** [comment]
Oui, et effectivement même si on ne détecte pas 100% des situations, j'ai l'impression que les deux cas cités (remplacements par `non` et variants mutuellement exclusifs) correspondent à 90% des avertissements dans https://github.com/betagouv/mon-entreprise/blob/c2149ac047278e1ac0be3befba4495a72a14275c/tmp. À voir si on peut aller au delà après.

**Comment by denismerigoux** [comment]
Cette issue est génial, tout est présenté de manière super claire. Du coup je confirme que c'est exactement ce que je traite avec Catala par la logique par défaut. Ce que je fait c'est que je lève systématiquement une erreur lorsque que les conditions de 2 remplacements pour la même variable sont vraies en même temps. Cependant, je garde la possibilité d'enlever cette erreur en donnant une "priorité" à un remplacement sur l'autre, priorité à spécifier dans le programme original (et à justifier par le texte de loi).

Impossible de faire une vérification statique facile, exhaustive et précise car ça nécessiterait de par exemple prouver que les conditions de deux remplacements sont mutuellement exclusives. Je prévois pour Catala de balancer ces obligations de preuve à un solveur SMT, on est obligé de sortir l'artillerie lourde.
